commit 563878ba217491dd0a6fbd588cd56d09e3456c14
Author: John Spray <jspray@redhat.com>
Date:   Wed Jul 26 07:44:00 2017 -0400

    mgr: refactor python interface
    
    Expose a python class instead of a module,
    so that we have a place to carry our reference
    to our MgrPyModule* and to PyModules*, rather than
    passing a handle for the former and using
    a global pointer for the latter.
    
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/MgrPyModule.cc b/src/mgr/MgrPyModule.cc
index 50165e71187..e3e3160de13 100644
--- a/src/mgr/MgrPyModule.cc
+++ b/src/mgr/MgrPyModule.cc
@@ -107,13 +107,31 @@ MgrPyModule::MgrPyModule(const std::string &module_name_, const std::string &sys
       PySys_SetObject(const_cast<char*>("stdout"), py_logger);
 #endif
     }
+
+    PySys_SetPath(const_cast<char*>(sys_path.c_str()));
+
     // Populate python namespace with callable hooks
-    Py_InitModule("ceph_state", CephStateMethods);
     Py_InitModule("ceph_osdmap", OSDMapMethods);
     Py_InitModule("ceph_osdmap_incremental", OSDMapIncrementalMethods);
     Py_InitModule("ceph_crushmap", CRUSHMapMethods);
 
-    PySys_SetPath(const_cast<char*>(sys_path.c_str()));
+    PyMethodDef ModuleMethods[] = {
+      {nullptr}
+    };
+
+    // Initialize module
+    PyObject *ceph_module = Py_InitModule("ceph_module", ModuleMethods);
+    assert(ceph_module != nullptr);
+
+    // Initialize base class
+    BaseMgrModuleType.tp_new = PyType_GenericNew;
+    if (PyType_Ready(&BaseMgrModuleType) < 0) {
+        assert(0);
+    }
+
+    Py_INCREF(&BaseMgrModuleType);
+    PyModule_AddObject(ceph_module, "BaseMgrModule",
+                       (PyObject *)&BaseMgrModuleType);
   }
 }
 
@@ -150,7 +168,7 @@ MgrPyModule::~MgrPyModule()
   }
 }
 
-int MgrPyModule::load()
+int MgrPyModule::load(PyModules *py_modules)
 {
   if (pMyThreadState.ts == nullptr) {
     derr << "No python sub-interpreter exists for module '" << module_name << "'" << dendl;
@@ -166,6 +184,8 @@ int MgrPyModule::load()
   if (pModule == nullptr) {
     derr << "Module not found: '" << module_name << "'" << dendl;
     derr << handle_pyerror() << dendl;
+
+    assert(0);
     return -ENOENT;
   }
 
@@ -178,15 +198,18 @@ int MgrPyModule::load()
     derr << handle_pyerror() << dendl;
     return -EINVAL;
   }
-
   
-  // Just using the module name as the handle, replace with a
-  // uuidish thing if needed
-  auto pyHandle = PyString_FromString(module_name.c_str());
-  auto pArgs = PyTuple_Pack(1, pyHandle);
+  // We tell the module how we name it, so that it can be consistent
+  // with us in logging etc.
+  
+  auto pThisPtr = PyCapsule_New(this, nullptr, nullptr);
+  auto pPyModules = PyCapsule_New(py_modules, nullptr, nullptr);
+  
+  auto pModuleName = PyString_FromString(module_name.c_str());
+  auto pArgs = PyTuple_Pack(3, pModuleName, pPyModules, pThisPtr);
   pClassInstance = PyObject_CallObject(pClass, pArgs);
   Py_DECREF(pClass);
-  Py_DECREF(pyHandle);
+  Py_DECREF(pModuleName);
   Py_DECREF(pArgs);
   if (pClassInstance == nullptr) {
     derr << "Failed to construct class in '" << module_name << "'" << dendl;
diff --git a/src/mgr/MgrPyModule.h b/src/mgr/MgrPyModule.h
index 9f5a5ef53ac..232d4ebac92 100644
--- a/src/mgr/MgrPyModule.h
+++ b/src/mgr/MgrPyModule.h
@@ -30,6 +30,7 @@
 
 
 class MgrPyModule;
+class PyModules;
 
 /**
  * A Ceph CLI command description provided from a Python module
@@ -60,7 +61,7 @@ public:
   MgrPyModule(const std::string &module_name, const std::string &sys_path, PyThreadState *main_ts);
   ~MgrPyModule();
 
-  int load();
+  int load(PyModules *py_modules);
   int serve();
   void shutdown();
   void notify(const std::string &notify_type, const std::string &notify_id);
